feat: experimental serialized configuration (#10791)
withastro/docs: src/content/docs/en/reference/experimental-flags/serialized-configuration.mdx

---
title: Experimental serialized configuration
sidebar:
  label: Serialized configuration
i18nReady: true
---

import Since from '~/components/Since.astro';

<p>

  **Type:** `boolean`<br />
  **Default:** `false`<br />

  <Since v="5.2.0" />
</p>

This feature allows to use a new virtual module called `astro:config`. The virtual module exposes a non-exhaustive, serializable, type-safe version of the Astro configuration.

The virtual module exposes two "submodules".

> ## `astro:config/client`
> 
> It exposes information that are **safe** to expose to the client (browser), hence it can be used in any instance of the application. It exposes information such as `trailingSlash`, `build.format`, `i18n`, and more.

Reviewer: ``````suggestion
### `astro:config/client`

The client submodule allows you to access a subset of your configuration values that are safe to expose to the browser.

#### Imports from `astro:config/client`
```js
import {
  i18n
  trailingSlash
  base
  build.format
  site
  redirects
 } from 'astro:actions';
 ```

These values can be used by any file in your Astro project:

```js title="src/utils.js"
import { trailingSlash } from "astro:config/client";

function addForwardSlash(path) {
  if (trailingSlash === "always") {
    return path.endsWith("/") ? path : path + "/"
  } else {
    return path
  }
}
```
``````
What about a structure like this for each of client and server submodules? This helps us prep for the "regular" model page by including the exports as its own section. (I pulled that list from the initial RFC, so we should update for whatever ended up in the current implementation!)

Note, either of those paragraphs (sentences!) could be beefed up if you want! Just thinking general structure here.

Reviewer: That's fair! Where is the most direct place we can link to in the RFC so people can clearly see which values are available from each of the two submodules? I'll update to write a note about experimental, updating constantly, blah blah blah, but would like the links to get people to these values as quickly as possible!

Author: Here https://github.com/withastro/roadmap/blob/feat/serialised-config/proposals/0051-serialized-manifest.md

> ## `astro:config/server`

Reviewer: NB: Did not make suggestions here, but whatever we do in the section above, we should duplicate here.

---
title: Experimental serialized configuration
sidebar:
  label: Serialized configuration
i18nReady: true
---

import Since from '~/components/Since.astro';

<p>

  **Type:** `boolean`<br />
  **Default:** `false`<br />

  <Since v="5.2.0" />
</p>

This feature allows access to the `astro:config` virtual module which exposes a non-exhaustive, serializable, type-safe version of the Astro configuration through two sub-modules.

For a complete overview, and to give feedback on this experimental API, see the [Serialized Manifest RFC](https://github.com/withastro/roadmap/blob/feat/serialised-config/proposals/0051-serialized-manifest.md).

To enable this virtual module, add the `experimental.serializeManifest` feature flag to your Astro config:

```js title="astro.config.mjs" ins={3-5}
import {defineConfig} from "astro/config"
export default defineConfig({
  experimental: {
    serializeManifest: true
  }
})
```

## Submodules

> The virtual module exposes two submodules for accessing different subsets of your configuration values. This protects your information by only making some data available to the client, while keeping other data available only on the server.

Reviewer: ```suggestion
The virtual module exposes two submodules for accessing different subsets of your configuration values. This protects your information by only making some data available to the client. 
```

## `astro:config/client`

> It exposes information that are **safe** to expose to the client (browser), hence it can be used in any instance of the application. It exposes information such as `trailingSlash`, `build.format`, `i18n`, and more.

Reviewer: ```suggestion
It exposes information that are **safe** to expose to the client (browser), hence it can be used in any instance of the application. It exposes information such as `trailingSlash`, `build.format`, `i18n`, and more.

This is a developing feature. For a full, up-to-date list of the configuration values available from `astro:config/client`, please see [the proposed API from the feature proposal](hhttps://github.com/withastro/roadmap/blob/feat/serialised-config/proposals/0051-serialized-manifest.md#proposed-apis:~:text=astro%3Aconfig%2Fclient%2C%20exported%20bindings%3A)
```

---
title: Experimental serialized configuration
sidebar:
  label: Serialized configuration
i18nReady: true
---

import Since from '~/components/Since.astro';

<p>

  **Type:** `boolean`<br />
  **Default:** `false`<br />

  <Since v="5.2.0" />
</p>

This feature allows access to the `astro:config` virtual module which exposes a non-exhaustive, serializable, type-safe version of the Astro configuration through two sub-modules.

For a complete overview, and to give feedback on this experimental API, see the [Serialized Manifest RFC](https://github.com/withastro/roadmap/blob/feat/serialised-config/proposals/0051-serialized-manifest.md).

To enable this virtual module, add the `experimental.serializeManifest` feature flag to your Astro config:

```js title="astro.config.mjs" ins={3-5}
import {defineConfig} from "astro/config"
export default defineConfig({
  experimental: {
    serializeManifest: true
  }
})
```

> 

Reviewer: ``````suggestion

Then, configuration values can be used by any file in your Astro project:
```js title="src/utils.js" ins={1}
import { trailingSlash } from "astro:config/client";
function addForwardSlash(path) {
  if (trailingSlash === "always") {
    return path.endsWith("/") ? path : path + "/"
  } else {
    return path
  }
}
```
``````
Realized we lost the example here!

## Submodules

The virtual module exposes two submodules for accessing different subsets of your configuration values. This protects your information by only making some data available to the client.

All available config values can be accessed from `astro:config/server`. However, for code executed on the client, only those values exposed by `astro:config/client` will be available.

## `astro:config/client`

The client submodule allows you to access a subset of the configuration values in `astro:config/server` that are safe to expose to the browser such as `trailingSlash`, `build.format`, `i18n`, and more. Use this submodule for client side code that is executed on the client.

This is a developing feature. For a full, up-to-date list of the configuration values available from `astro:config/client`, please see [the proposed API from the feature RFC](https://github.com/withastro/roadmap/blob/feat/serialised-config/proposals/0051-serialized-manifest.md#astroconfigclient)

## `astro:config/server`

> The server submodule allows you to access non-exhaustive set of your configuration values from `astro.config.mjs`. This includes `astro:config/client` values such as `trailingSlash` and `i18n`, but also more sensitive information about your file system configuration that is not safe to expose to the client such as `srcDir`, `cacheDir`, `outDir`.

Reviewer: ```suggestion
The server submodule allows you to access a non-exhaustive set of your configuration values from `astro.config.mjs`. This includes `astro:config/client` values such as `trailingSlash` and `i18n`, but also more sensitive information about your file system configuration that is not safe to expose to the client such as `srcDir`, `cacheDir`, `outDir`.
```

For a complete overview, and to give feedback on this experimental API, see the [Serialized Manifest RFC](https://github.com/withastro/roadmap/blob/feat/serialised-config/proposals/0051-serialized-manifest.md).

To enable this virtual module, add the `experimental.serializeManifest` feature flag to your Astro config:

```js title="astro.config.mjs" ins={3-5}
import {defineConfig} from "astro/config"
export default defineConfig({
  experimental: {
    serializeManifest: true
  }
})
```

Then, configuration values can be used by any file in your Astro project:
```js title="src/utils.js" ins={1}
import { trailingSlash } from "astro:config/client";
function addForwardSlash(path) {
  if (trailingSlash === "always") {
    return path.endsWith("/") ? path : path + "/"
  } else {
    return path
  }
}
```
## Submodules

The virtual module exposes two submodules for accessing different subsets of your configuration values. This protects your information by only making some data available to the client.

All available config values can be accessed from `astro:config/server`. However, for code executed on the client, only those values exposed by `astro:config/client` will be available.

> ## `astro:config/client`
> 
> The client submodule allows you to access a subset of the configuration values in `astro:config/server` that are safe to expose to the browser such as `trailingSlash`, `build.format`, `i18n`, and more. Use this submodule for client side code that is executed on the client.
> 
> This is a developing feature. For a full, up-to-date list of the configuration values available from `astro:config/client`, please see [the proposed API from the feature RFC](https://github.com/withastro/roadmap/blob/feat/serialised-config/proposals/0051-serialized-manifest.md#astroconfigclient)
> 
> ## `astro:config/server`

Reviewer: ```suggestion
### `astro:config/client`

The client submodule allows you to access a subset of the configuration values in `astro:config/server` that are safe to expose to the browser such as `trailingSlash`, `build.format`, `i18n`, and more. Use this submodule for client side code that is executed on the client.

This is a developing feature. For a full, up-to-date list of the configuration values available from `astro:config/client`, please see [the proposed API from the feature RFC](https://github.com/withastro/roadmap/blob/feat/serialised-config/proposals/0051-serialized-manifest.md#astroconfigclient)

### `astro:config/server`
```
